File: dp_labs_and_solutions/dp_labs_and_solutions/dp_matlab_labs_and_solutions/command/lab/DpVector.m
% DpVector.m

classdef DpVector < DpContainer
	properties (GetAccess = private, SetAccess = private)
		NumberOfElements % int
		Elements % cell array of NumberOfElements elements
	end

	methods
		% constructor:
		function theDpVector = DpVector
			theDpVector.NumberOfElements = 0;
			theDpVector.Elements = cell(0, 1);
		end

		function add(theDpVector, aValue)
			newElements = cell(theDpVector.NumberOfElements + 1, 1);
			for i = 1:theDpVector.NumberOfElements
				newElements{i} = theDpVector.Elements{i};
			end
			newElements{theDpVector.NumberOfElements + 1} = aValue;
			theDpVector.NumberOfElements = theDpVector.NumberOfElements + 1;
			theDpVector.Elements = newElements;
		end

		function set(theDpVector, aSubscript, aValue)
			theDpVector.validateSubscript(aSubscript);
			theDpVector.Elements{aSubscript} = aValue;
		end

		function result = get(theDpVector, aSubscript)
			theDpVector.validateSubscript(aSubscript);
			result = theDpVector.Elements{aSubscript};
		end

		function result = makeIterator(theDpVector)
			result = DpVectorIterator(theDpVector);
		end

		function result = makeReverseIterator(theDpVector)
			result = DpVectorReverseIterator(theDpVector);
		end

		function result = getSize(theDpVector)
			result = theDpVector.NumberOfElements;
		end
	end

	methods (Access = private)
		function validateSubscript(theDpVector, aSubscript)
			if (aSubscript < 1) || (aSubscript > theDpVector.getSize)
				error('Invalid aSubscript.');
			end
		end
	end
end

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/clone/solution/DpVector.java
// DpVector.java

import java.util.Vector;

public class DpVector<T> extends DpContainer<T>
{
	public DpVector()
	{
		myElements = new Vector<T>();
	}

	public void add(T aValue)
	{
		myElements.add(aValue);
	}

	public int getSize()
	{
		return myElements.size();
	}

	public void set(int aSubscript, T aValue)
	{
		validateSubscript(aSubscript);

		myElements.set(aSubscript, aValue);
	}

	public T get(int aSubscript)
	{
		validateSubscript(aSubscript);

		return myElements.get(aSubscript);
	}

	public DpIterator<T> makeIterator()
	{
		return new DpVectorIterator<T>(this);
	}

	public DpIterator<T> makeReverseIterator()
	{
		return new DpVectorReverseIterator<T>(this);
	}

	private void validateSubscript(int aSubscript)
	{
		if ((aSubscript < 0) || (aSubscript >= getSize()))
		{
			throw new DpException("Invalid aSubscript.");
		}
	}

	private Vector<T> myElements;

	private static class DpVectorIterator<T> extends DpIterator<T>
	{
		public DpVectorIterator(DpVector<T> aVector)
		{
			myVector = aVector;
			myIndex  = 0;
		}

		public boolean isDone()
		{
			return (myIndex >= myVector.getSize());
		}

		public T getCurrent()
		{
			return myVector.get(myIndex);
		}

		public void advance()
		{
			myIndex++;
		}

		private DpVector<T> myVector;
		int myIndex;
	}

	private static class DpVectorReverseIterator<T> extends DpIterator<T>
	{
		public DpVectorReverseIterator(DpVector<T> aVector)
		{
			myVector = aVector;
			myIndex = aVector.getSize() - 1;
		}

		public boolean isDone()
		{
			return (myIndex < 0);
		}

		public T getCurrent()
		{
			return myVector.get(myIndex);
		}

		public void advance()
		{
			myIndex--;
		}

		private DpVector<T> myVector;
		private int myIndex;
	}
}
